wgsl/extract-grammar.py

```diff
@@ -510,7 +761,321 @@ def read_spec(options):
+            elif s_v == "?":
+                functools.reduce(
+                    lambda x, y: x["members"][y],
+                    scan_in["content"][0:-1],
+                    syntax_dict[scan_in["value"]],
+                )["members"][-1] = {
+                    "type": "match_0_1",
+                    "members": [
+                        functools.reduce(
+                            lambda x, y: x["members"][y],
+                            scan_in["content"][0:-1],
+                            syntax_dict[scan_in["value"]],
+                        )["members"][-1]
+                    ],
+                }
+                s_i += 1
+            elif s_v == "+":
+                functools.reduce(
+                    lambda x, y: x["members"][y],
+                    scan_in["content"][0:-1],
+                    syntax_dict[scan_in["value"]],
+                )["members"][-1] = {
+                    "type": "match_1_n",
+                    "members": [
+                        functools.reduce(
+                            lambda x, y: x["members"][y],
+                            scan_in["content"][0:-1],
+                            syntax_dict[scan_in["value"]],
+                        )["members"][-1]
+                    ],
+                }
+                s_i += 1
+            elif s_v == "*":
+                functools.reduce(
+                    lambda x, y: x["members"][y],
+                    scan_in["content"][0:-1],
+                    syntax_dict[scan_in["value"]],
+                )["members"][-1] = {
+                    "type": "match_0_n",
+                    "members": [
+                        functools.reduce(
+                            lambda x, y: x["members"][y],
+                            scan_in["content"][0:-1],
+                            syntax_dict[scan_in["value"]],
+                        )["members"][-1]
+                    ],
+                }
+                s_i += 1
+            elif s_v == "|":
+                functools.reduce(
+                    lambda x, y: x["members"][y],
+                    scan_in["content"][0:-2],
+                    syntax_dict[scan_in["value"]],
+                )["members"][-1]["type"] = "choice"
+                s_i += 1
+            elif s_v == "(":
+                functools.reduce(
+                    lambda x, y: x["members"][y],
+                    scan_in["content"][0:-1],
+                    syntax_dict[scan_in["value"]],
+                )["members"].append({"type": "sequence", "members": []})
+                scan_in["content"].append(0)
+                s_i += 1
+            elif s_v == ")":
+                scan_in["content"] = scan_in["content"][0:-1]
+                s_i += 1
+            else:
+                s_i += 1
+        else:
+            s_i += 1
+        if s_i < len(bnfdialect_data):
+            s_v = bnfdialect_data[s_i]
+        else:
+            s_v = ""
+
+    def reproduce_rule(production):
```

Comment:
This should be the __str__ method on a rule class.
I assume this returns a string?